File: web/src/jsMain/kotlin/io/vitalir/web/common/Loadable.kt
package io.vitalir.web.common

sealed interface Loadable<out T> {

    val valueOrNull: T?
        get() = if (this is Loaded) value else null

    object Loading : Loadable<Nothing>
    data class Loaded<T>(val value: T) : Loadable<T>
    data class Error(val exception: Exception) : Loadable<Nothing>

}

File: web/src/core/models/Loadable.ts
export interface Loading {
    kind: "loading"
}
export interface Loaded<T> {
    kind: "loaded"
    data: T
}
export interface LoadingError {
    kind: "error"
    error: string
}

export type Loadable<T> = Loading | Loaded<T> | LoadingError
